Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp

```diff
@@ -163,14 +163,22 @@ std::optional<Color> BitmapImageDescriptor::singlePixelSolidColor() const
     return primaryNativeImageMetadata(m_singlePixelSolidColor, std::optional<Color>(), CachedFlag::SinglePixelSolidColor, &NativeImage::singlePixelSolidColor);
 }
 
-Headroom BitmapImageDescriptor::headroom() const
+bool BitmapImageDescriptor::hasHDRGainMap() const
 {
-    return primaryImageFrameMetadata(m_headroom, CachedFlag::Headroom, &ImageFrame::headroom);
+    return imageMetadata(m_hasHDRGainMap, false, CachedFlag::HasHDRGainMap, &ImageDecoder::hasHDRGainMap);
 }
 
-bool BitmapImageDescriptor::hasHDRGainMap() const
+bool BitmapImageDescriptor::hasHDRColorSpace() const
 {
-    return imageMetadata(m_hasHDRGainMap, false, CachedFlag::HasHDRGainMap, &ImageDecoder::hasHDRGainMap);
+    if (m_cachedFlags.contains(CachedFlag::ColorSpace))
+        return m_colorSpace.usesITUR_2100TF();
+
+    if (m_source->primaryNativeImageIfExists())
+        return colorSpace().usesITUR_2100TF();
+
+    bool hasHDRContent = colorSpace().usesITUR_2100TF();
+    m_source->destroyNativeImageAtIndex(m_source->primaryFrameIndex());
```

Comment:
There's an implicit assumption here that reading `colorSpace()` allocated the frame (because that's what `NativeImageSource` does. But that seems like an implementation detail of `NativeImageSource `, so this frame destruction should move there.

Reply:
Destroying the frame here is unnecessary. But we need to fix the async image decoding code path such that it ignores this frame and treat it as if it did not exist. 

It seems ImageIO caches the frame also. So if we call `CGImageSourceCreateImageAtIndex()` ImageIO will return the cached frame without having to create a new one.

I added a FIXME comment and I will address fixing the async image decoding issue in a separate PR.